src/coreclr/src/tools/Common/JitInterface/SystemVStructClassificator.cs

```diff
@@ -12,9 +12,9 @@ namespace Internal.JitInterface
     using static SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR;
     using static SystemVClassificationType;
 
-    internal class SystemVStructClassificator
+    internal static class SystemVStructClassificator
     {
-        private Dictionary<TypeDesc, SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR> _classificationCache = new Dictionary<TypeDesc, SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR>();
+        private static Dictionary<TypeDesc, SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR> s_classificationCache;
```

Comment:
We don't make statics that refer to type system entities (or NodeCache nodes) in the compiler. Doing so limits reusability of the JitInterface. We already have use cases where one can have multiple type system contexts in the same process: in those cases, static cache is just another word for "memory leak". I would prefer this stays an instance field (it also avoids the problems with races Sergiy pointed out).

Reply:
But that means that in multi-threaded compilation, we will need to run the classification for the same type upto as many times as there are compilation threads and we would have multiple copies of the classification structure. And I would also need one more classifier for the ArgIterator that doesn't seem to be bound to any JitInterface.
So the idea of adding the classification to the TypeDesc and removing the cache seems to be much better.

Reply:
It's a compromise. We try to keep TypeDesc as small as possible because there's _a lot_ of them.

SystemV classification is only needed on Unix and only for types that are used as method arguments. Is it worth burning `sizeof(SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR)` (aligned) bytes on each TypeDesc (or a pointer to the struct, whichever is smaller) to avoid recomputing the same information ~9 times?

Reply:
Actually, for the ArgIterator, it is worse, as we create an instance for every function call to generate GCRefMap. So we would repeat the classification over and over and also possibly create a lot of memory garbage unless we can somehow get the current thread's CorInfoImpl (or propagate the current CorInfoImpl down to the place where we create the ArgIterator instance). I'm not sure how reasonable that would be.

Reply:
> we create an instance for every function call

We don't have means to pass an extra argument with the classificator to gcrefmap computation?

I'm okay adding this to TypeDesc if the math checks out. But since we only need this information for

* value types,
* used as arguments,
* and on Unix only

my gut feel is that side cache is better.

Reply:
Thanks for the feedback, I'll look into passing the classifier to the gcrefmap computation. And I am actually also running an experiment at the moment to check what would happen to the build time of all the pri1 tests and frame work assemblies together if there is no caching of that stuff at all. That might also be an option if it turns out to be causing only a minor difference.

Reply:
So, disabling the caching completely results in virtually zero difference in compilation time of all the framework assemblies and pri1 tests. In fact, the run without the caching was slightly faster, which I assume is just noise.
```
With the cache:
Total build time: 5962387 msecs
Framework time:   133722 msecs
Compilation time: 4017379 msecs
Execution time:   1811279 msecs

Without the cache:
Total build time: 5944179 msecs
Framework time:   133107 msecs
Compilation time: 3981134 msecs
Execution time:   1829931 msecs
```
So I am going to remove the cache.